release/default.nix

```diff
@@ -99,34 +99,53 @@ let
     }
   ];
 
-  modifiedPkgNames = attrNames (import ../pkgs/overlay.nix pkgs pkgs);
-
+  # Recursively searches a package attrset for derivations and returns a list
+  # of attribute paths as "dotted names", like "adns" or "fc.agent".
+  getDottedPackageNames = attrs: prevPath:
+    filter
+      (p: p != null)
+      (lib.flatten
+        ((lib.mapAttrsToList
+          (n: v:
+            let
+              path = prevPath ++ [n];
+              dottedName = (lib.concatStringsSep "." path);
+              shouldRecurse = (isAttrs v && v.recurseForDerivations or false);
+            in
+              if lib.isDerivation v then dottedName
+              else if shouldRecurse then getDottedPackageNames v path
+              else null)
+          attrs)));
+
+  # Exclude packages from being built by Hydra.
+  # The exclusion list is applied to overlay packages and important packages.
+  # Supports excluding packages from nested sets using "dotted names" like "fc.blockdev".
   excludedPkgNames = [
-    # The kernel universe is _huge_ and contains a lot of unfree stuff. Kernel
-    # packages which are really needed are pulled in as dependencies anyway.
-    "linux"
-    "linux_5_4"
-    "linuxPackages"
-    "linuxPackages_5_4"
-    # XXX: fails on 21.05, must be fixed
-    "backy"
-    # This is a store path in the overlay which doesn't work in Hydra restricted mode.
-    # We don't need to test it.
-    "gitlab-runner"
   ];
 
-  includedPkgNames = [
-    "awscli"
-    "calibre"
-    "element-web"
-    "firefox"
-    "gitlab-ee"
-  ];
+  overlay = import ../pkgs/overlay.nix pkgs pkgs;
+  overlayPkgNames = getDottedPackageNames overlay [];
+  overlayPkgNamesToTest = lib.subtractLists excludedPkgNames overlayPkgNames;
+
+  importantPkgNames = fromJSON (readFile ../important_packages.json);
+  importantPkgNamesToTest = lib.subtractLists excludedPkgNames importantPkgNames;
 
-  testPkgNames = includedPkgNames ++
-    lib.subtractLists excludedPkgNames modifiedPkgNames;
+  # Results looks like: [ { python3Packages.requests.x86_64-linux = <job>; } ]
+  pkgNameToHydraJobs = dottedName:
+    let
+      path = lib.splitString "." dottedName;
+      job = lib.hydraJob (lib.attrByPath path null pkgs);
+    in
+      map
+        (system: lib.setAttrByPath (path ++ [ system ]) job)
+        supportedSystems;
 
-  testPkgs = listToAttrs (map (n: { name = n; value = pkgs.${n}; }) testPkgNames);
+  pkgNameListToHydraJobs = pkgNameList:
+    # Merge the single-attribute sets from pkgNameToHydraJobs into one big attrset.
+    lib.foldl'
```

Comment:
strict version because it is still small enough to be evaluated on our small hydra machines?

> The difference is that evaluation is forced upon access. Usually used
with small whole results (in contrast with lazily-generated list or large
lists where only a part is consumed.)

Reply:
`foldl'` is typically used in similar functions in nixpkgs and it's the builtin function for that. I don't think it even makes a difference here because everything will be evaluated anyway. The strict variant is probably better for error messages.